main.go

```diff
@@ -1,31 +1,105 @@
 package main
 
 import (
-	"fmt"
-	"github.com/vinegarhq/merlin/internal"
+	"encoding/csv"
+	"encoding/json"
+	"flag"
 	"log"
+	"net/http"
 	"os"
-)
 
-func usage() {
-	fmt.Fprintln(os.Stderr, "usage: merlin [path to configuration JSON]")
-	os.Exit(1)
-}
+	"golang.org/x/time/rate"
+)
 
 func main() {
-	if len(os.Args) < 2 {
-		usage()
+	cf := flag.String("config", "", "Path to configuration file")
+	flag.Parse()
+
+	if *cf == "" {
+		log.Fatal("cannot run without configuration file")
 	}
 
-	config, err := internal.LoadConfiguration(os.Args[1])
+	cfg, err := LoadConfig(*cf)
 	if err != nil {
 		log.Fatal(err)
 	}
-	log.Println("Loaded configuration successfully.")
 
-	log.Println("Starting HTTP listener...")
-	err = internal.BeginListener(config)
+	log.Fatal(serve(&cfg))
+}
+
+func serve(cfg *Config) error {
+	f, err := os.OpenFile(cfg.OutputFile, os.O_WRONLY|os.O_APPEND|os.O_CREATE, 0o600)
 	if err != nil {
-		log.Fatal(err)
+		return err
+	}
+	defer f.Close()
+
+	fi, err := f.Stat()
+	if err != nil {
+		return err
+	}
+
+	csvw := csv.NewWriter(f)
+
+	if fi.Size() < 1 {
+		if err := csvw.Write(CSVHeader); err != nil {
+			return err
+		}
+
+		csvw.Flush()
+	} else {
+		log.Println("Warning: will not write CSV header to existing output file")
 	}
+
+	limiter := rate.NewLimiter(rate.Limit(cfg.RateLimit), 1)
+
+	http.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
```

Comment:
wrap `func(w http.ResponseWriter, req *http.request)` with tbLimiter, or it will not work. 
e.g.:

`http.HandleFunc("/", tb.LimitFuncHandler(tbLimiter, func(w http.ResponseWriter, r *http.Request)`

Reply:
Check 3 lines below.